Data/models/RecentSearches.swift

```diff
@@ -0,0 +1,97 @@
+// Copyright 2021 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+import CoreData
+import Shared
+
+private let log = Logger.browserLogger
+
+public enum RecentSearchType: Int32 {
+    case qrCode = 0
+    case text = 1
+    case website = 2
+}
+
+@objc(RecentSearch)
+final public class RecentSearch: NSManagedObject, CRUD {
+    @NSManaged public var searchType: Int32
+    @NSManaged public var text: String?
+    @NSManaged public var websiteUrl: String?
+    @NSManaged public var dateAdded: Date?
+    
+    public class func frc() -> NSFetchedResultsController<RecentSearch> {
+        let context = DataController.viewContext
+        let fetchRequest = NSFetchRequest<RecentSearch>()
+        fetchRequest.entity = RecentSearch.entity(context)
+        fetchRequest.fetchBatchSize = 5
+        
+        let createdSort = NSSortDescriptor(key: "dateAdded", ascending: false)
+        fetchRequest.sortDescriptors = [createdSort]
+        
+        return NSFetchedResultsController(fetchRequest: fetchRequest, managedObjectContext: context,
+                                          sectionNameKeyPath: nil, cacheName: nil)
+    }
+    
+    public static func addItem(type: RecentSearchType, text: String?, websiteUrl: String?) {
+        DataController.perform(context: .new(inMemory: false), save: true) { context in
+            let item = RecentSearch(context: context)
+            item.searchType = type.rawValue
+            item.text = text
+            item.websiteUrl = websiteUrl
+            item.dateAdded = Date()
+        }
+    }
+    
+    public static func getItem(text: String) -> RecentSearch? {
+        return RecentSearch.first(where: NSPredicate(format: "text == %@", text))
+    }
+    
+    public static func itemExists(text: String) -> Bool {
+        return getItem(text: text) != nil
```

Comment:
Regarding the count vs first comment, we can add a `exists(where: predicate)` method to `CRUDProtocols` that would check for count > 0, it gives us performance gain and it's easy to read and use